### PR TITLE
fix(diff): don't report byte-different files as identical

### LINKED ISSUE #3469: rtk diff reports "Files are identical" (exit 0) for byte-different files
# `rtk diff` reports "Files are identical" (exit 0) for byte-different files

## Summary

`rtk diff` (the built-in "ultra-condensed diff") gives **false-equal** results in two
independent ways, both dangerous when `diff` is used to *prove* two files are equal:

1. **Whitespace/line-ending normalization → false "identical".** Files whose only
   differences are trailing whitespace, CRLF-vs-LF, a missing/extra trailing newline,
   or leading indentation are reported as `[ok] Files are identical` even though the
   files are **not** byte-identical (different md5, and `cmp`/`diff` report a
   difference).

2. **Exit code is always 0.** Even when `rtk diff` *does* detect and display a real
   content change, it exits `0`. GNU/BSD `diff` exits `1` when files differ. This
   breaks `if diff a b; then ...` and `diff a b && echo same`.

Either defect silently defeats any verification of the form "diff proves these two
files are equal."

## Environment

- `rtk 0.42.4` (the token-killer `rtk`; `rtk gain` works — not the unrelated
  `reachingforthejack/rtk`). Current Homebrew stable is `0.44.2` — the behavior
  below should be re-verified against the latest release in case it is already
  addressed.
- Installed via Homebrew
- macOS (Darwin 25.5.0), Apple Silicon
- Triggered in practice through a Claude Code `PreToolUse` hook (`rtk hook claude`)
  that rewrites bare `diff a b` → `rtk diff a b`, but the defects reproduce by
  calling `rtk diff` directly (see below), so the hook is not required.

### Defect 1 — whitespace-only difference reported as identical

```bash
printf 'foo\nbar\n'  > A      # md5: f47c75614087a8dd938ba4acff252494
printf 'foo \nbar\n' > B      # md5: 259f776914a6f16abb525d25d8c84aa1  (trailing space)

md5 A B            # -> two DIFFERENT hashes
cmp A B; echo $?   # -> "A B differ: char 4, line 1"   exit 1
diff A B; echo $?  # (real diff) -> "1c1 < foo / > foo "   exit 1

rtk diff A B; echo $?
# ACTUAL:   [ok] Files are identical
#           0
# EXPECTED: a reported difference, non-zero exit
```

Same false "identical" reproduces for:

| Case | file A | file B | md5 A | md5 B |
|------|--------|--------|-------|-------|
| trailing space | `foo\nbar\n` | `foo \nbar\n` | f47c7561… | 259f7769… |
| CRLF vs LF | `alpha\nbeta\n` | `alpha\r\nbeta\r\n` | 852e77b4… | def13413… |
| trailing newline | `abc` | `abc\n` | 90015098… | 0bee89b0… |
| indentation | `x\n    indented\n` | `x\nindented\n` | 08990ce7… | 8a4dbe59… |

All four: `rtk diff` → `[ok] Files are identical`, exit `0`.
Ground truth for all four: different md5, `cmp` exit `1`, real `diff` exit `1`.

### Defect 2 — exit code 0 even when a real change is shown

```bash
printf 'a\nworld\n' > A
printf 'a\nWORLD\n' > B

rtk diff A B; echo $?
# ACTUAL:
#   A → B
#      +1 added, -1 removed, ~0 modified
#   -   2 world
#   +   2 WORLD
#   0            <-- exit 0 despite a displayed difference
# EXPECTED: exit 1
```

## Impact

`diff` is commonly used as an equality oracle in scripts and deploy verifications
(`diff a b && echo "unchanged"`, `if diff old new; then skip; fi`). Both defects turn
a genuine difference into a "they match" verdict. The whitespace case is especially
risky because whitespace/line-ending changes (e.g. a file silently converted to CRLF,
or a lost trailing newline) are exactly the kind of drift a verification is meant to
catch.

## Suggested fix

- `rtk diff` should exit **non-zero when the files differ** (match `diff`'s convention:
  0 = same, 1 = differ, 2 = error), on both the "identical" path and the
  changes-displayed path.
- Whitespace/line-ending/trailing-newline differences should **not** be collapsed to
  "identical" by default. If line-normalized comparison is desirable, gate it behind an
  explicit opt-in flag (e.g. `-w` / `--ignore-whitespace`) rather than making it the
  default, and even then report a non-zero exit for byte-level differences.
- At minimum, when the byte contents differ but the normalized view is equal, print a
  distinct message (e.g. "files differ only in whitespace/line endings") and exit
  non-zero — never `[ok] Files are identical`.

## Workaround

Add `diff` to `exclude_commands` in `config.toml` so the hook passes it through to the
real binary:

```toml
[hooks]
exclude_commands = ["diff"]
```

Verified: with this set, `rtk`-hooked `diff` runs the system `diff` and correctly
reports differences with exit 1 for all cases above. `rtk proxy diff a b` also bypasses
the filter. For equality proofs we now standardize on `cmp`/`md5` (which have no `rtk`
subcommand and pass through untouched) rather than `diff`.

### PR DESCRIPTION
**Pull request by ousamabenyounes**
## Summary

- `rtk diff` could report `[ok] Files are identical` with exit 0 for byte-different files whose only differences are hidden by `str::lines()`, such as CRLF-vs-LF or a trailing newline (#3469).
- Fix: pass the original file contents into the renderer and only use the identical branch when `content1 == content2`. If the line diff is empty but the file contents differ, `rtk diff` now prints an explicit whitespace/line-ending message and exits 1.
- This is rebased onto current `develop` and keeps the newer `select_file_diff_output` path, so the empty-diff message bypasses the tiny-output `never_worse` fallback that swallowed the reviewer-requested small-file case.
- Added Hydr0gen19's byte-identical CRLF guard test from #3471, adapted to this PR's current renderer signature with no fifth `render_file_diff`/renderer argument.

## Test verification (RED -> GREEN)

New coverage:

- Unit tests for CRLF-vs-LF, trailing-newline-only, and byte-identical CRLF content.
- Run-level integration test for a small CRLF-vs-LF file pair through `rtk diff`, covering the review feedback path.

**RED on unmodified `origin/develop`**

```
assertion `left == right` failed: [ok] Files are identical

  left: Some(0)
 right: Some(1)
test result: FAILED. 0 passed; 1 failed
```

**GREEN on this branch**

```
test small_crlf_vs_lf_diff_prints_whitespace_message ... ok
test result: ok. 1 passed; 0 failed
```

Local CI replay:

- Baseline `run-ci.sh` on `origin/develop`: test-presence, fmt, clippy, and `cargo test --all` passed; unit suite `2929 passed; 0 failed; 8 ignored`.
- Final `run-ci.sh` on this branch: test-presence, fmt, clippy, and `cargo test --all` passed; unit suite `2932 passed; 0 failed; 8 ignored`; new integration test `1 passed`.
- Coverage: `cargo llvm-cov --all` passed, and the Rust changed-line LCOV check found covered executable changed lines only (`changed executable lines with coverage data: 80`, PASS).

## Test plan

- [x] `cargo fmt --all -- --check`
- [x] `cargo clippy --all-targets`
- [x] `cargo test --all`
- [x] `cargo llvm-cov --lcov --output-path ... --all`
- [x] RED proof on unmodified `origin/develop` for the new run-level regression test

> All PRs target the `develop` branch.

Fixes #3469

### CODE REVIEW

File: src/cmds/git/diff_cmd.rs [hunk 1]
```diff
@@ -40,7 +40,22 @@ fn render_file_diff(file1: &Path, file2: &Path, content1: &str, content2: &str)
...
+        // `str::lines()` strips `\r` and drops a trailing newline, so CRLF-vs-LF
+        // and trailing-newline-only differences leave no line-level change.
+        // Claiming "identical" for byte-different files is a false-equal result
+        // that defeats `rtk diff` used to verify equality (issue #3469). Report
+        // identical only when the bytes truly match; otherwise surface the
+        // whitespace/line-ending difference and exit 1 per the diff convention.
+        if content1 == content2 {
+            return ("[ok] Files are identical\n".to_string(), 0);
+        }
+        let msg = format!(
```
Comment (KuSh) [inline src/cmds/git/diff_cmd.rs]: This message is ~92 bytes (≈23 estimated tokens) plus both paths, and `run()` prints `never_worse(&raw, &rtk)` where `raw = format!("{}\n---\n{}", content1, content2)`. So whenever the two files together are under roughly 90 bytes, the guard discards this message and prints the raw concatenation instead — which is precisely the small-file case #3469 reports.

Measured on your own build:

```
$ rtk diff /tmp/lf.txt /tmp/crlf.txt      # "alpha\nbeta\n" vs "alpha\r\nbeta\r\n"
alpha
beta

---
alpha
beta
exit=1                                    # message never printed

$ rtk diff big_lf.txt big_crlf.txt        # same content, 5 longer lines
/tmp/big_lf.txt → /tmp/big_crlf.txt
   files differ only in whitespace or line endings (no line-content change)
exit=1
```

Two ways out, either is fine:

1. Shorten the message enough to survive the guard, e.g. `"{} != {}: whitespace/EOL only\n"`.
2. Return early past `never_worse` for this branch — the raw fallback is actively misleading here, since it renders two byte-different files as two identical-looking blocks.

I'd lean towards (2) with a short comment saying why, because (1) leaves the behaviour size-dependent and someone will lengthen the string later.

Either way this needs a test through `run()` rather than `render_file_diff()`. Both new unit tests call `render_file_diff` directly, so they pass regardless of what the guard does — the bug is invisible to them.

**Review by KuSh — CHANGES_REQUESTED** [review]
Requesting changes on one point. The fix itself is right and I want it in — this is about it actually reaching the user.

**Verified:** `diff.changes.is_empty()` is equivalent to "the `lines()` vectors are element-wise equal", so the only byte differences that can survive that check are exactly `\r\n`-vs-`\n` and a trailing newline. The guard's message is therefore accurate, not a guess, and the exit code correctly becomes 1. `cargo fmt --check`, `cargo clippy --all-targets` and the 25 `diff_cmd` tests all pass on your head. I also ran the binary end-to-end against the cases in #3469.

**The problem:** for small files the message never gets printed. Details inline.

Worth knowing, since it explains why this is worth another round: I confirmed on 0.46.0 that the *pre-existing* behaviour already prints a raw dump for these inputs —

```
$ printf 'a\r\nb\r\nc\r\n' > x; printf 'a\nb\nc\n' > y
$ rtk diff x y; echo "exit=$?"
a
b
c

---
a
b
c
exit=0
```

— so a user comparing two small files today sees two visually identical blobs and exit 0. Your PR fixes the exit code half of that. If the message is also swallowed, they get two visually identical blobs and exit 1, which is better but still leaves them to work out *why* on their own. That's the gap worth closing while you're in here.

**Comment by KuSh** [comment]
Also, since @Hydr0gen19 has already approved it, could you include the `test_render_byte_identical_exit_zero_with_crlf` test from #3471 in your PR?

**Comment by Hydr0gen19** [comment]
Yes, please take it — copy it straight across, no attribution needed.

One adjustment when you do: the version on #3471 passes a fifth argument to `render_file_diff` (`false`), which belongs to an `--ignore-whitespace` flag that this PR rightly does not carry. Drop that argument and it compiles against your signature unchanged. Your guard sits inside the `diff.changes.is_empty()` branch rather than ahead of the split, but the case still reaches your identical path, so the assertions hold either way.

Why it is worth having next to your two: both of yours pin the "must differ" direction. This one pins the other side — two files that are both CRLF and byte-identical must still report identical and exit 0. A guard that over-triggers whenever it sees a CRLF would pass `test_render_crlf_vs_lf_not_identical` and `test_render_trailing_newline_not_identical`, and fail only this one.

**Comment by ousamabenyounes** [comment]
Thanks @KuSh and @Hydr0gen19. Rebased onto current `develop` and pushed `f4325bc`: the whitespace/EOL-only branch now uses the existing empty-diff output path to bypass the size guard, so small CRLF/LF files print the explanatory message and exit 1 instead of being swallowed into `[ok] Files are identical`.

I also included Hydr0gen19's byte-identical CRLF guard test from #3471, adapted to this PR's current renderer signature with no fifth `render_file_diff`/renderer argument.

Validation: the new run-level regression is RED on `origin/develop` (`[ok] Files are identical`, exit 0) and GREEN here. Local `run-ci.sh` baseline/final passed test-presence, fmt, clippy, and `cargo test --all` (`2929 -> 2932` unit tests, plus the new integration test). `cargo llvm-cov` and the Rust changed-line coverage check passed.

Remote CI on `f4325bc`: test presence, fmt, doc review, clippy, Ubuntu/macOS/Windows tests, security scan, semgrep, and CLA passed. `benchmark` failed on the existing `golangci-lint` empty-output filter; the current `develop` run `33541678720` fails the same benchmark step with the same `BENCHMARK FAILED: 1 filter(s) returned empty output` summary.

**Review by KuSh — APPROVED** [review]
The message reaches the user now — clearing my earlier request for changes.

Rebasing onto `develop` was the right move and it solved this structurally rather than by special-casing. `select_file_diff_output` returns the rendered output unguarded when `changes.is_empty()`, which is exactly the whitespace branch, so the message survives at any file size instead of only above ~90 bytes:

```console
$ rtk diff lf.txt crlf.txt          # 11 bytes each — previously printed a raw dump
lf.txt → crlf.txt
   files differ only in whitespace or line endings (no line-content change)
exit=1
```

And `tests/diff_byte_accuracy_test.rs` is the run()-level test I asked for — it spawns the real binary on the *small* fixture, which is precisely the case the unit tests were blind to.

### Verified

`cargo fmt --all --check` and `cargo clippy --all-targets` clean, 3016 tests passing. Exit codes match `/usr/bin/diff` across the edge cases, and each branch is genuinely reachable:

| | verdict | exit (rtk / GNU) |
|---|---|---|
| `a\n` vs `a` | whitespace message | 1 / 1 |
| CRLF vs LF | whitespace message | 1 / 1 |
| `\n` vs `\r\n` | whitespace message | 1 / 1 |
| `a\r\n` vs `a\r\r\n` | real diff | 1 / 1 |
| `a\r` vs `a` | real diff | 1 / 1 |
| byte-identical | identical | 0 / 0 |

The last two matter: they confirm the message isn't over-claiming. `str::lines()` loses exactly two things — whether the final line ends in `\n`, and whether a `\r` preceded each `\n` — and both are line terminators, so "differ only in whitespace or line endings" is exact by construction rather than a best guess.

I also mutation-checked the new tests rather than trusting that they pass:

- byte check → line check (the original bug): both new unit tests fail.
- whitespace branch returning 0 instead of 1: both fail.
- removing the guard bypass so the message is swallowed again: `small_crlf_vs_lf_diff_prints_whitespace_message` fails with empty stdout.

That third one is the important one — the regression I flagged is now genuinely pinned, not passing by accident.

### One small change pushed

I pushed `428af61` on top: `render_diff` was taking `content1` and `content2` only to compare them, which gave a rendering function five parameters and an input comparison to do. The caller already holds both strings, so it now passes `content1 == content2` as a `bytes_equal` verdict. Behaviour is identical — same tests, same output, same exit codes, re-verified after the change. Revert it if you'd rather keep the contents in the signature.
